### DISCUSSION TITLE
Poetry project as local dependency

### DISCUSSION THREAD

**Discussion by Tegomena** [python-poetry/poetry#4092] [General] [answered]
Hi,

I'm quite new to Poetry, so please be patient ;-)

I'm developing a library in Poetry which will be published on PyPi later. In addition, I have some example scripts which are using my Poetry developed library. But these files are in another folder, outside the Poetry project as I don't want to publish them was part of the library. But of course they need to import the new library already and I have no clue how I can achieve this: scripts in a folder a with an import like this:

import MYLIBRARY

where MYLIBRARY is a Peotry project in another folder b somewhere else? First publishing the lib to PyPi and then installing it would work, but during development, I don't want to publsih permanently ...

I'm quite sure there is a simple solution for this, but I can't find it ;-)

**Comment by abn** [comment, marked as answer]
Here is an example that hopefully demonstrates what you are attempting to do. I am using the latest preview release for this. But should work for latest stable as well.

#### Create new project
I am using the source layout here since its cleaner, and prevents package imports working by virtue of python adding CWD to your PYTHONPATH.

```console
$ poetry new --src foobar
Created package foobar in foobar
$ tree foobar/
foobar/
├── pyproject.toml
├── README.md
├── src
│   └── foobar
│       └── __init__.py
└── tests
    └── __init__.py

3 directories, 4 files
```

#### Create example script
```console
$ cd foobar/
$ mkdir -p contrib/examples
$ cat > contrib/examples/hello.py <<EOF
> if __name__ == '__main__':
    import foobar
    print(foobar)
> EOF
```

#### Ensure virtualenv is created and project installed
This is the step where poetry will install your project as an editable package to your project virtual environment.
```console
$ poetry install
Updating dependencies
Resolving dependencies... (0.1s)

Writing lock file

Installing the current project: foobar (0.1.0)
```

#### Run sample script
```console
$ poetry run python contrib/examples/hello.py 
<module 'foobar' from '/tmp/foobar/src/foobar/__init__.py'>
```

Here, the `poetry run` command activates your virtual environment then executes `python contrib/examples/hello.py`. Since the project virutal environment is activated, the `python` executable form the environment with the corresponding python site and PYTHONPATH is used.

Alternatively you can also use `poetry shell` to first spawn a sub-shell with the environment activated. You can also activate the environment manually using `source .venv/bin/activate` (this assumes you have `poetry config virtualenvs.in-project true` configured - creating the project virtual environment within the project directory).

**Reply by Tegomena** [reply]
Ah, thanks a lot for this great explanation, I missed the "poetry install" part :-)
